File: python/getting_started/searching/Authenticator.py
import base64
import hmac
import hashlib
from datetime import datetime
from time import mktime
from wsgiref.handlers import format_date_time

class Authenticator(object):

    def __init__(self, app_id, app_secret, algorithm='hmac-sha1', digestmod=hashlib.sha1):
        self.app_id = app_id
        self.app_secret = app_secret
        self.algorithm = algorithm
        self.digestmod = digestmod

    def __call__(self, request):
        # body = request.body
        # if body and not isinstance(body, bytes):   # Python 3
        #     body = body.encode('utf-8')  # standard encoding for HTTP

        request_dt = mktime(datetime.now().timetuple())
        request_dt = format_date_time(request_dt)
        signing_base = f'path: {request.path_url}\ndate: {request_dt}'

        signature = hmac.new(
            self.app_secret,
            signing_base.encode('utf-8'),
            digestmod=self.digestmod).digest()
        signature = base64.b64encode(signature).decode('utf-8')

        request.headers['Date'] = request_dt
        request.headers['Path'] = request.path_url
        request.headers['Authorization'] = f'hmac username="{self.app_id}",algorithm="{self.algorithm}",headers="path date",signature="{signature}"'

        return request

File: netcore/hcos.getting_started/searching/Authenticator.cs
using RestSharp;
using System;
using System.Linq;
using System.Text;
using System.Security.Cryptography;
using RestSharp.Authenticators;
namespace hcos.getting_started.searching
{
    /// <summary>
    /// HMAC implementation of Restsharp IAuthenticator.
    /// </summary>
    public class Authenticator : IAuthenticator
    {
        /// <summary>
        /// Application name.
        /// </summary>
        public string AppId { get; set; }

        /// <summary>
        /// Application secret.
        /// </summary>
        public string AppSecret { get; set; }
        /// <summary>
        /// Tenant Id
        /// </summary>
        public string TenantId { get; set; }

        /// <summary>
        /// Contructor initialization using appName and secret.
        /// </summary>
        /// <param name="appName"></param>
        /// <param name="secret"></param>
        public Authenticator(string appName, string appSecret, string tenantId)
        {
            AppId = appName;
            AppSecret = appSecret;
            TenantId = tenantId;
        }
        #region RestSharp.IAuthenticator
        /// <summary>
        /// Method implementing point of entry for RestSharp authentication.
        /// </summary>
        /// <param name="client"></param>
        /// <param name="request"></param>
        public void Authenticate(IRestClient client, IRestRequest request)
        {
            DateTime requestDT = DateTime.Now.ToUniversalTime();
            Uri finalUri = client.BuildUri(request);
            string algorithm = "hmac-sha1";

            string signingBase = $"path: {finalUri.PathAndQuery}\ndate: {requestDT.ToString("R")}";

            byte[] secretBytes = Encoding.ASCII.GetBytes(AppSecret);

            string signature;

            using (var hmac = new HMACSHA1(secretBytes))
            {
                var hash = hmac.ComputeHash(Encoding.UTF8.GetBytes(signingBase));
                signature = Convert.ToBase64String(hash);
            }

            request.AddOrUpdateParameter("Date", requestDT.ToString("R"), ParameterType.HttpHeader);
            request.AddOrUpdateParameter("Path", finalUri.PathAndQuery, ParameterType.HttpHeader);
            request.AddOrUpdateParameter("Authorization", $"hmac username=\"{AppId}\",algorithm=\"{algorithm}\",headers=\"path date\",signature=\"{signature}\"", ParameterType.HttpHeader);
        }
        #endregion
    }
}